common/functions/src/aggregates/aggregate_retention.rs

```diff
@@ -0,0 +1,230 @@
+            if new_columns[j as usize].get_data(row) {
+                state.add(j);
+            }
+        }
+
+        Ok(())
+    }
+
+    fn serialize(&self, place: StateAddr, writer: &mut BytesMut) -> Result<()> {
+        let state = place.get::<AggregateRetentionState>();
+        state.serialize(writer)
+    }
+
+    fn deserialize(&self, place: StateAddr, reader: &mut &[u8]) -> Result<()> {
+        let state = place.get::<AggregateRetentionState>();
+        state.deserialize(reader)
+    }
+
+    fn merge(&self, place: StateAddr, rhs: StateAddr) -> Result<()> {
+        let rhs = rhs.get::<AggregateRetentionState>();
+        let state = place.get::<AggregateRetentionState>();
+        state.merge(rhs);
+        Ok(())
+    }
+
+    #[allow(unused_mut)]
+    fn merge_result(
+        &self,
+        place: StateAddr,
+        array: &mut dyn common_datavalues::MutableColumn,
+    ) -> Result<()> {
+        let state = place.get::<AggregateRetentionState>();
+
+        let builder: &mut MutableObjectColumn<JsonValue> = Series::check_get_mutable_column(array)?;
+        let mut vec: Vec<JsonValue> = Vec::new();
+
+        let mut flag: bool = false;
+
+        for i in 0..self.events_size {
+            if flag || state.events & (1 << i) == 0 {
+                vec.push(json!(0));
+                if i == 0 {
+                    flag = true;
+                }
+            } else {
+                vec.push(json!(1));
+            }
+        }
+        builder.append_value(JsonValue::Array(vec));
+
+        Ok(())
+    }
+
+    fn accumulate_keys(
+        &self,
+        places: &[StateAddr],
+        offset: usize,
+        columns: &[common_datavalues::ColumnRef],
+        _input_rows: usize,
+    ) -> Result<()> {
+        for (row, place) in places.iter().enumerate() {
+            self.accumulate_row(place.next(offset), columns, row)?;
+        }
+        Ok(())
+    }
+}
+
+impl fmt::Display for AggregateRetentionFunction {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self.display_name)
+    }
+}
+
+impl AggregateRetentionFunction {
+    pub fn try_create(
+        display_name: &str,
+        arguments: Vec<DataField>,
+    ) -> Result<AggregateFunctionRef> {
+        Ok(Arc::new(Self {
+            display_name: display_name.to_owned(),
+            events_size: arguments.len() as u8,
+            _arguments: arguments,
+        }))
+    }
+}
+
+pub fn try_create_aggregate_retention_function(
+    display_name: &str,
+    _params: Vec<DataValue>,
+    arguments: Vec<DataField>,
+) -> Result<AggregateFunctionRef> {
+    assert_variadic_arguments(display_name, arguments.len(), (1, 32))?;
+
+    for argument in arguments.iter() {
+        let data_type = argument.data_type();
+        if data_type.data_type_id() != TypeID::Boolean {
+            return Err(ErrorCode::BadDataValueType(format!(
+                "AggregateRetention does not support type '{:?}'",
```

Comment:
So this function expects which type?

Reply:
This function expect BooleanTyp. So we should print it out.  It looks better.

Reply:
Sorry, this expression should expect expression.